community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/QueryPlanner.scala

```diff
@@ -32,19 +33,42 @@ trait QueryPlanner {
 
 case class DefaultQueryPlanner(planRewriter: Rewriter,
                                config: QueryPlannerConfiguration = QueryPlannerConfiguration.default,
-                               planSingleQuery: LogicalPlanningFunction1[PlannerQuery, LogicalPlan] = planSingleQueryX())
+                               planSingleQuery: LogicalPlanningFunction1[PlannerQuery, LogicalPlan] = PlanSingleQuery())
   extends QueryPlanner {
 
   def plan(unionQuery: UnionQuery)(implicit context: LogicalPlanningContext): LogicalPlan = unionQuery match {
     case UnionQuery(queries, distinct) =>
-      val plan = planQuery(queries, distinct)
-      plan.endoRewrite(planRewriter)
+      val plan = planQueries(queries, distinct)
+      val rewrittenPlan = plan.endoRewrite(planRewriter)
+      createProduceResultOperator(rewrittenPlan, unionQuery)
 
     case _ =>
       throw new CantHandleQueryException
   }
 
-  private def planQuery(queries: Seq[PlannerQuery], distinct: Boolean)(implicit context: LogicalPlanningContext) = {
+  private def createProduceResultOperator(in: LogicalPlan, unionQuery: UnionQuery)
+                                         (implicit context: LogicalPlanningContext): LogicalPlan = {
+    val nodes = mutable.ListBuffer[String]()
+    val rels = mutable.ListBuffer[String]()
+    val others = mutable.ListBuffer[String]()
+
+    val lastQuery = unionQuery.queries.last
+    val columns = lastQuery.lastQueryHorizon.exposedSymbols(lastQuery.lastQueryGraph)
+
+    columns.foreach {
```

Comment:
Maybe it is better to add convenience methods `isNode(IdName)` and `isRelationship(IdName)` to `SemanticTable` and replace this foreach with mutable state with smth like:
```
val nodes = columns.filter(c => context.semanticTable.isNode(c))
val rels = columns.filter(c => context.semanticTable.isRelationship(c))
val others = columns.filter(c => !context.semanticTable.isNode(c) && !context.semanticTable.isRelationship(c))
```